### ISSUE TITLE
Yolov8 training with apple m2 pro silicion mps - slow training 

### ISSUE THREAD

**Issue by fbgencer** [ultralytics/ultralytics#5717]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I am trying to speed up training with Mac mini m2 pro but I believe there is a problem.

I followed this example (https://docs.ultralytics.com/modes/train/#apple-m1-and-m2-mps-training)
```py
from ultralytics import YOLO
# Load a model
model = YOLO('yolov8n.pt')  # load a pretrained model (recommended for training)
# Train the model with 2 GPUs
results = model.train(data='coco128.yaml', epochs=3, imgsz=640, device='mps')
```

Considering suggestions of @glenn-jocher about this issue https://github.com/ultralytics/ultralytics/issues/4695, I tried to run with 3 epochs for each device "cpu" and "mps", but "cpu" is faster than "mps" which is unexpected. For inference on yolov8x model I'm getting almost 33ms duration which covers my expectations but in training feels like there is a bug. Additionally, my device has 16gb ram and also active cooling so I think it can handle a little bit load.

- Setting device to "cpu", running the above lines

```
optimizer: AdamW(lr=0.000119, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs/detect/train3
Starting training for 3 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        1/3         0G      1.096      1.365      1.202        201        640: 100%|██████████| 8/8 [00:45<00:00,  5.72s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:19<00:00,  4.84s/it]
                   all        128        929      0.658      0.532      0.615      0.457

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        2/3         0G      1.216      1.443      1.268        136        640: 100%|██████████| 8/8 [00:45<00:00,  5.69s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:19<00:00,  4.85s/it]
                   all        128        929      0.675      0.539      0.628      0.467

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        3/3         0G      1.192      1.342      1.242        206        640: 100%|██████████| 8/8 [00:45<00:00,  5.69s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:19<00:00,  4.86s/it]
                   all        128        929      0.685      0.536      0.632       0.47

3 epochs completed in 0.054 hours.
Optimizer stripped from runs/detect/train3/weights/last.pt, 6.5MB
Optimizer stripped from runs/detect/train3/weights/best.pt, 6.5MB

Validating runs/detect/train3/weights/best.pt...
Ultralytics YOLOv8.0.198 🚀 Python-3.9.18 torch-2.1.0 CPU (Apple M2 Pro)
```

- Setting device to "mps",

```
Logging results to runs/detect/train5
Starting training for 3 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        1/3         0G       1.23      3.574      1.251        201        640: 100%|██████████| 8/8 [01:16<00:00,  9.62s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [05:49<00:                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [05:49<00:00, 87.46s/it]
                   all        128        929      0.485      0.399      0.408      0.269

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        2/3         0G      1.375      3.482      1.349        136        640: 100%|██████████| 8/8 [01:33<00:00, 11.71s/
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [03:44<00:
                   all        128        929      0.491      0.394       0.42       0.28

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        3/3         0G      1.357      3.155      1.299        206        640: 100%|██████████| 8/8 [01:56<00:00, 14.51s/
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [03:11<00:
                   all        128        929      0.487      0.397      0.419      0.276

3 epochs completed in 0.295 hours.
Optimizer stripped from runs/detect/train5/weights/last.pt, 6.5MB
Optimizer stripped from runs/detect/train5/weights/best.pt, 6.5MB

Validating runs/detect/train5/weights/best.pt...
Ultralytics YOLOv8.0.198 🚀 Python-3.9.18 torch-2.1.0 MPS (Apple M2 Pro)
```

Are these results expected? Is this a known bug? Thank you for such a good project 👏

### Environment

Python version
```
 $ python
Python 3.9.18 (main, Sep 11 2023, 08:25:10) 
[Clang 14.0.6 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```
Pip list results
```
 $ pip list                                                                                                        127 ↵
Package             Version
------------------- --------------------
certifi             2023.7.22
charset-normalizer  3.3.0
contourpy           1.1.1
cycler              0.12.1
filelock            3.12.4
fonttools           4.43.1
fsspec              2023.9.2
idna                3.4
importlib-resources 6.1.0
Jinja2              3.1.2
kiwisolver          1.4.5
MarkupSafe          2.1.3
matplotlib          3.8.0
mpmath              1.3.0
networkx            3.1
numpy               1.26.1
opencv-python       4.8.1.78
packaging           23.2
pandas              2.1.1
Pillow              10.0.1
pip                 23.2.1
psutil              5.9.5
py-cpuinfo          9.0.0
pyparsing           3.1.1
python-dateutil     2.8.2
pytz                2023.3.post1
PyYAML              6.0.1
requests            2.31.0
scipy               1.11.3
seaborn             0.13.0
setuptools          68.0.0
six                 1.16.0
sympy               1.12
thop                0.1.1.post2209072238
torch               2.1.0
torchvision         0.16.0
tqdm                4.66.1
typing_extensions   4.8.0
tzdata              2023.3
ultralytics         8.0.198
urllib3             2.0.6
wheel               0.41.2
zipp                3.17.0
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by freds-dev** [comment]
I am experiencing the same thing. With `device` set to `mps` it takes about 10x longer than with the default.
Any updates on this? @fbgencer

**Comment by glenn-jocher** [comment]
@freds-dev it seems you're encountering performance issues when setting `device` to `mps`. This may be due to the `mps` (multi-process service) mode relying on the sharing of GPU resources among multiple processes which can sometimes lead to slower performance, especially if the GPU resources are not efficiently allocated.

An update that may improve processing time is under development but not yet released.

In the meantime, if performance is critical for your use case, you might want to consider using the default device setting or experiment with other configurations that could optimize the use of your system's resources.

Remember to keep an eye on the repository for updates and improvements to the software. Thank you for raising this issue - your feedback helps us to improve the performance of YOLOv8.

**Comment by fbgencer** [comment]
After a close examination on training code I think I found a clue.

When we call model.train, it uses multiple inherited classes and base classes to achieve the desired functionality. In the trainer code I measured the time spent on specific functions and realized that loss calculations on mps takes too much time. Such that, as in my first message, calculation on cpu takes 0.2secs, while on mps takes 8secs. 
Here training first apply preprocess on batches, creates loss values first using model forward and then calculation loss on the predictions: https://github.com/ultralytics/ultralytics/blob/465df3024f44fa97d4fad9986530d5a13cdabdca/ultralytics/engine/trainer.py#L340-L345
After looking the model.__call__ or forward function we get https://github.com/ultralytics/ultralytics/blob/465df3024f44fa97d4fad9986530d5a13cdabdca/ultralytics/nn/tasks.py#L35-L45
calls self.loss 

https://github.com/ultralytics/ultralytics/blob/465df3024f44fa97d4fad9986530d5a13cdabdca/ultralytics/nn/tasks.py#L200-L215
inits criterion here

https://github.com/ultralytics/ultralytics/blob/465df3024f44fa97d4fad9986530d5a13cdabdca/ultralytics/nn/tasks.py#L290-L293
then 
https://github.com/ultralytics/ultralytics/blob/465df3024f44fa97d4fad9986530d5a13cdabdca/ultralytics/utils/loss.py#L164-L208

For the last function there are different calculations on the loss and each of them takes around 1secs if the tensor lies in mps. So I know it is not a feasible solution, I converted mps tensors to cpu and back again after loss calculation is done. The time takes for conversion is much less than the mps-loss calculation (probably depends on too many factors i.e memory-bandwith m1 or m2 chip etc)

Now my code is still dirty and couldn't write a wrapper, but above lines describe my progress. Since in the first epoch model warms up, after that I am getting very fast calculations such as getting 6secs (per epoch) for yolov8n model and dataset in my first post.

However, in the validation phase of the training I still couldn't figure it out what the problem is but working on it.

When I try to clean up I am planning the share the codes, maybe my progress gives an idea to the dev team.

**Comment by glenn-jocher** [comment]
Thank you for the detailed analysis and your contribution towards improving YOLOv8's performance, @fbgencer. 

You've identified that the loss calculations under the `mps` device setting seem to be a point of inefficiency. It's particularly insightful that you've indicated the significant disparity in calculation times between using CPU and `mps` in this context. 

The approach you've used to convert `mps` tensors to CPU and back again after loss calculation is clever, and as per your observation, could help reduce overall computation time despite the conversion cost due to minimized loss calculation time.

Regarding the validation phase issue you've mentioned, it might also be a matter of how operations are performed under different device modes, just like what you've observed during the training phase.

As you continue refining your solution and exploring the validation phase issue, any additional information you share would indeed be helpful. Your insights and progress stand to greatly benefit the development team and the entire user community in optimizing YOLOv8's performance.

We appreciate your valuable input and encourage you to keep us updated on your continuing work. Your active contribution is instrumental in making YOLOv8 better for everyone.

**Comment by twilson1111** [comment]
Checking in - the mps speed seems to be better on my M2, but now the training doesn't converge for my dataset when using mps.  Has anyone else run into the same issue?  I am trying to figure out training parameters (batch size, learning rate, optimizer) that help but no luck so far.

**Comment by glenn-jocher** [comment]
@twilson1111 it's great to hear that you're seeing an improvement in speed on your M2 using `mps`! 🚀 Regarding the convergence issue, you're definitely on the right track by tweaking the training parameters. Changing the optimizer to Adam and adjusting the learning rate are excellent steps. Every dataset is unique, so finding the right combination can require some experimentation. 

For those encountering similar issues, here's a quick snippet on how to adjust these settings:

```python
from ultralytics import YOLO

# Load the model, specifying device as 'mps'
model = YOLO('yolov8n.pt', device='mps')

# Training with adjusted parameters
results = model.train(data='your_dataset.yaml', epochs=100, lr0=0.00025, optimizer='Adam')
```

It seems like smaller learning rates are beneficial in some scenarios, especially when training on distinct or smaller datasets. If anyone else is experiencing similar challenges, consider adjusting your learning rate or experimenting with different optimizers. 🛠️

Keep us posted on your progress, and if you discover more insights or effective configurations, sharing them would be immensely helpful to the community!
